src/Functions/array/arrayFold.cpp

```diff
@@ -0,0 +1,257 @@
+
+    const size_t arguments_count = arguments.size();
+    assert(arguments_count > 2);
+    ColumnPtr column_offset;
+    const ColumnArray * column_first_array = nullptr;
+    ColumnsWithTypeAndName arrays;
+    arrays.reserve(arguments_count - 1);
+
+    for (size_t i = 1; i < arguments_count - 1; ++i) 
+    {
+        const auto & array_with_type_and_name = arguments[i];
+        ColumnPtr column_array_ptr = array_with_type_and_name.column;
+        const auto * column_array = checkAndGetColumn<ColumnArray>(column_array_ptr.get());
+        if (!column_array)
+        {
+            const ColumnConst * column_const_array = checkAndGetColumnConst<ColumnArray>(column_array_ptr.get());
+            if (!column_const_array)
+                throw Exception(
+                    ErrorCodes::ILLEGAL_COLUMN, "Expected array column, found {}", column_array_ptr->getName());
+            column_array_ptr = recursiveRemoveLowCardinality(column_const_array->convertToFullColumn());
+            column_array = checkAndGetColumn<ColumnArray>(column_array_ptr.get());
+        }
+
+        const DataTypePtr & type_array_ptr = array_with_type_and_name.type;
+        const auto * type_array = checkAndGetDataType<DataTypeArray>(type_array_ptr.get());
+        if (!type_array)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Expected array type, found {}.", type_array_ptr->getName());
+
+        if (!column_offset)
+        {
+            column_offset = column_array->getOffsetsPtr();
+        }
+        else
+        {
+            /// The first condition is optimization: do not compare data if the pointers are equal.
+            if (column_array->getOffsetsPtr() != column_offset
+                && column_array->getOffsets() != typeid_cast<const ColumnArray::ColumnOffsets &>(*column_offset).getData())
+                throw Exception(
+                    ErrorCodes::BAD_ARGUMENTS, "Arrays passed to {} must have equal size.", getName());
+        }
+        if (i == 1)
+        {
+            column_first_array = column_array;
+        }
+        arrays.emplace_back(ColumnWithTypeAndName(column_array->getDataPtr(),
+                                                  recursiveRemoveLowCardinality(type_array->getNestedType()),
+                                                  array_with_type_and_name.name));
+    }
+    arrays.emplace_back(arguments.back());
+
+    if (rows_count == 0)
+            return arguments.back().column->convertToFullColumnIfConst()->cloneEmpty();
+
+    ColumnWithTypeAndName column_accumulator = arguments.back();
+    std::vector<size_t> array_size_vec(rows_count);
+    const ColumnArray::Offsets& array_offsets = column_first_array->getOffsets();
+    size_t max_array_size = rows_count ? array_offsets[0] : 0;
+    std::map<size_t, ColumnPtr> row_res;
+
+    for (size_t i = 0; i < rows_count; ++i)
+    {
+        auto size = array_offsets[i] - array_offsets[i - 1];
```

Comment:
I found i - 1 = -1 when i = 0 by reading the code.
I don't know if it's relate to your core dump

Reply:
Thanks four your reply. 

 I think -1 here is not the reason of core dump for 2 reasons:
1. core dump happend before this line, the number of this line is 175, and core happened in line 169  with code: `const IColumn::Offsets& array_offsets = column_first_array->getOffsets();`
2. -1 is valid. Because array_offsets has type `PaddedPODArray<UInt64>`, which receives a signed index to access -1th element without pointer overflow. So array_offsets[-1] will return 0 as expected.

Reply:
Because the type of offsets is `WrappedPtr` which uses counter ref. After the loop, `ColumnPtr column_array_ptr` is destructed, then the offsets disappear as well. I'm not sure but gdb tells me offsets ptr of `column_first_array` is null out of the loop but is not null inside the loop, so I think it's the only reason.

if so, two solutions:
1. use `ColumnPtr column_first_array`
2. use `column_offset`

If not, let me know

Reply:
seems it's late night in China, so I fix it for you.
Feel free to change if you have other ideas. Happy weekend!

Reply:
Thank you, that's very kind of you! 

And you are right, `ColumnPtr` will be destructed after loop, this is exactly the reason of the previous core dump. BUT, i still have some questions that why core dump does not happen every time when i call arrayFold.

Also, have a good weekend!